File: 1331/1331.py
'''
Fraction = '0' | ['-'] UnsignedFraction

not LL(1)
UnsignedFraction = WholePart ' ' Numerator '/' Denominator |
    Numerator '/' Denominator |
    WholePart

WholePart = PositiveInt
Numerator = PositiveInt
Denominator = PositiveInt

LL(1)
UnsignedFraction = PositiveInt [' ' PositiveInt '/' PositiveInt | '/' PositiveInt]  # [] - 0 or 1, if

PositiveInt = '1'..'9' {'0'..'9'}  # {} - 0 or more times, while

5 6/7
5/6
5
'''

from math import gcd
from typing import NamedTuple

s = ''
pos = 0
ch = ''
EOT = '\n'

def nextChar():
    global ch, pos

    if pos < len(s):
        ch = s[pos]
        pos += 1
    else:
        ch = EOT

def readLine():
    global s, pos

    s = input()
    pos = 0

    nextChar()

# PositiveInt = '1'..'9' {'0'..'9'}
def positiveInt():
    assert '1' <= ch <= '9'
    v = ord(ch) - ord('0')
    nextChar()

    while '0' <= ch <= '9':
        d = ord(ch) - ord('0')
        nextChar()
        v = v * 10 + d

    return v

class Fraction(NamedTuple):
    num: int  # numerator
    den: int  # denominator

    def __add__(self, right):
        return Fraction(self.num * right.den + right.num * self.den, self.den * right.den)  # 5/6 + 7/8 = (5 * 8) / (6 * 8) + (7 * 6) / (6 * 8) = (5 * 8 + 7 * 6) / (6 * 8)

    def __sub__(self, right):
        return Fraction(self.num * right.den - right.num * self.den, self.den * right.den)  # 5/6 - 7/8 = (5 * 8) / (6 * 8) - (7 * 6) / (6 * 8) = (5 * 8 - 7 * 6) / (6 * 8)

    def __mul__(self, right):
        return Fraction(self.num * right.num, self.den * right.den)  # 5/6 * 7/8 = (5 * 7) / (6 * 8)

    def __truediv__(self, right):
        new_num = self.num * right.den
        new_den = self.den * right.num

        assert new_den != 0

        if new_den > 0:
            return Fraction(new_num, new_den)
        else:
            return Fraction(-new_num, -new_den)  # -5 / -3 = 5 / 3, 5 / -3 = -5 / 3

# UnsignedFraction = PositiveInt [' ' PositiveInt '/' PositiveInt | '/' PositiveInt]
def unsignedFraction():
    v = positiveInt()

    if ch == ' ':
        nextChar()
        num = positiveInt()
        assert ch == '/'
        nextChar()
        den = positiveInt()
        return Fraction(v * den + num, den)  # 5 6/7 = 5 * 7 / 7 + 6/7 = (5 * 7 + 6) / 7

    elif ch == '/':
        nextChar()
        den = positiveInt()
        return Fraction(v, den)
    else:
        return Fraction(v, 1)

# Fraction = '0' | ['-'] UnsignedFraction
def fraction():
    if ch == '0':
        nextChar()
        return Fraction(0, 1)
    else:
        sign = 1

        if ch == '-':
            sign = -1
            nextChar()

        ans = unsignedFraction()
        return Fraction(ans.num * sign, ans.den)

def readFraction():
    readLine()
    ans = fraction()
    assert ch == EOT

    return ans

def printFraction(f: Fraction):
    num = f.num
    den = f.den

    assert den > 0

    if num < 0:
        print('-', end='')
        num = -num

    assert num >= 0

    if num % den == 0:
        print(num // den)
    else:
        whole = num // den
        num %= den
        assert num < den

        if whole != 0:
            print(f'{whole} ', end='')

        g = gcd(num, den)
        assert g >= 1

        num //= g
        den //= g

        assert den > 1
        print(f'{num}/{den}')  # print("%d/%d" % (num, den))

a = readFraction()
sign = input()
b = readFraction()

if sign == '+':
    printFraction(a + b)
elif sign == '-':
    printFraction(a - b)
elif sign == '*':
    printFraction(a * b)
else:
    assert sign == '/'
    printFraction(a / b)

File: 1331/1331.cpp
#include <cstdio>
#include <cassert>
#include <string>
#include <numeric>

struct Fraction {
    // long long до 2^63-1 ~= 9e18
    long long num; // numerator
    long long den; // denominator
    // den > 0
};

Fraction operator +(const Fraction &left, const Fraction &right) {
    // 2/3 + 1/5 = (2 * 5) / (3 * 5) + (1 * 3) / (5 * 3) = (2 * 5 + 1 * 3) / (3 * 5)
    long long num = left.num * right.den + right.num * left.den;
    long long den = left.den * right.den;

    return Fraction{num, den};
}

Fraction operator -(const Fraction &left, const Fraction &right) {
    // 2/3 - 1/5 = (2 * 5) / (3 * 5) - (1 * 3) / (5 * 3) = (2 * 5 - 1 * 3) / (3 * 5)
    long long num = left.num * right.den - right.num * left.den;
    long long den = left.den * right.den;

    return Fraction{num, den};
}

Fraction operator *(const Fraction &left, const Fraction &right) {
    // 2/3 * 1/5 = (2 * 1) / (3 * 5)
    long long num = left.num * right.num;
    long long den = left.den * right.den;

    return Fraction{num, den};
}

Fraction operator /(const Fraction &left, const Fraction &right) {
    // (2/3) / (1/5) = (2 * 5) / (3 * 1)
    // (-2/3) / (1/5) = (-2 * 5) / (3 * 1)
    // (2/3) / (-1/5) = (2 * 5) / (3 * -1) = - (2 * 5) / (3 * 1)
    // (-2/3) / (-1/5) = (-2 * 5) / (3 * -1) = - (-2 * 5) / (3 * 1)

    long long num = left.num * right.den;
    long long den = left.den * right.num;

    if (den < 0) {
        den = -den;
        num = -num;
    }

    return Fraction{num, den};
}

Fraction readFraction() {
    char buffer[18 + 1] = {};
    char endl = '\n';
    scanf("%18[^\n]%c", buffer, &endl);
    assert(endl == '\n');
    std::string s(buffer);

    int spacePos = s.find(' ');
    assert(spacePos != 0);

    if (spacePos >= 0) {
        int wholePart = std::stoi(s.substr(0, spacePos)); // int(s[:spacePos])
        std::string fracPart = s.substr(spacePos + 1);

        int slashPos = fracPart.find('/');
        assert(slashPos >= 0);

        int num = std::stoi(fracPart.substr(0, slashPos));
        int den = std::stoi(fracPart.substr(slashPos + 1));
        // 2 3/5 -> (2 * 5 + 3) / 5 -> 13/5
        // -2 3/5 -> (-2 * 5 - 3) / 5 -> -13/5

        if (wholePart < 0) {
            return Fraction{wholePart * den - num, den};
        } else {
            return Fraction{wholePart * den + num, den};
        }
    } else {
        int slashPos = s.find('/');

        if (slashPos >= 0) {
            int num = std::stoi(s.substr(0, slashPos));
            int den = std::stoi(s.substr(slashPos + 1));

            return Fraction{num, den};
        } else {
            int num = std::stoi(s);
            return Fraction{num, 1};
        }
    }
}

void printFraction(Fraction f) {
    assert(f.den > 0);

    if (f.num < 0) {
        printf("-");
        f.num = -f.num;
    }

    assert(f.num >= 0);

    if (f.num % f.den == 0) {
        printf("%lld", f.num / f.den);
    } else {
        long long whole = f.num / f.den;
        f.num %= f.den;
        assert(f.num < f.den);

        if (whole != 0) {
            printf("%lld ", whole);
        }

        long long g = std::gcd(f.num, f.den);
        f.num /= g;
        f.den /= g;

        assert(f.den > 1);
        printf("%lld/%lld", f.num, f.den);
    }
}

int main() {
    Fraction a = readFraction();

    char endl = '\n';
    char sign = '?';
    scanf("%c%c", &sign, &endl);
    assert(endl == '\n');

    Fraction b = readFraction();

    if (sign == '+') {
        printFraction(a + b);
    } else if (sign == '-') {
        printFraction(a - b);
    } else if (sign == '*') {
        printFraction(a * b);
    } else {
        assert(sign == '/');
        printFraction(a / b);
    }

    return 0;
}
